services/s3/bucket_delete.go

```diff
@@ -0,0 +1,218 @@
+			return objCount, fmt.Errorf("listing S3 bucket (%s) object versions: %w", bucket, err)
+		}
+
+		n, err := fn(ctx, c.client, bucket, page)
+		objCount += n
+
+		if err != nil {
+			lastErr = err
+			break
+		}
+	}
+
+	if lastErr != nil {
+		return objCount, lastErr
+	}
+
+	return objCount, nil
+}
+
+func boolPtr(b bool) *bool {
+	return &b
+}
+
+func getObjectsToDelete(page *s3.ListObjectVersionsOutput) []s3.ObjectIdentifier {
+	if page.Versions == nil {
+		return nil
+	}
+
+	toDelete := make([]s3.ObjectIdentifier, 0, len(*page.Versions))
+	for _, v := range *page.Versions {
+		toDelete = append(toDelete, s3.ObjectIdentifier{
+			Key:       v.Key,
+			VersionId: v.VersionId,
+		})
+	}
+
+	return toDelete
+}
+
+func getDeleteMarkersToDelete(page *s3.ListObjectVersionsOutput) []s3.ObjectIdentifier {
+	if page.DeleteMarkers == nil {
+		return nil
+	}
+
+	toDelete := make([]s3.ObjectIdentifier, 0, len(*page.DeleteMarkers))
+	for _, v := range *page.DeleteMarkers {
+		toDelete = append(toDelete, s3.ObjectIdentifier{
+			Key:       v.Key,
+			VersionId: v.VersionId,
+		})
+	}
+
+	return toDelete
+}
+
+func deletePageOfObjectVersions(ctx context.Context, conn *s3.APIClient, bucket string, force bool, page *s3.ListObjectVersionsOutput) (int64, error) {
+	toDelete := getObjectsToDelete(page)
+	var objCount int64
+	if objCount = int64(len(toDelete)); objCount == 0 {
+		return objCount, nil
+	}
+
+	req := conn.ObjectsApi.DeleteObjects(ctx, bucket).DeleteObjectsRequest(s3.DeleteObjectsRequest{
+		Objects: &toDelete,
+		Quiet:   boolPtr(true),
+	})
+	if force {
+		req = req.XAmzBypassGovernanceRetention(true)
+	}
+
+	output, apiResponse, err := req.Execute()
+	if apiResponse.HttpNotFound() {
+		return objCount, nil
+	}
+
+	if err != nil {
+		return objCount, fmt.Errorf("deleting S3 bucket (%s) object versions: %w", bucket, err)
+	}
+
+	if output.Errors == nil {
+		return objCount, nil
+	}
+
+	objCount -= int64(len(*output.Errors))
+	var errs []error
+	for _, v := range *output.Errors {
+		if force && toString(v.Code) == errAccessDenied {
+			_, err := tryDisableLegalHold(ctx, conn, bucket, toString(v.Key), toString(v.VersionId))
+			if err != nil {
+				errs = append(errs, []error{newDeleteObjectVersionError(v), fmt.Errorf("removing legal hold: %w", newObjectVersionError(toString(v.Key), toString(v.VersionId), err))}...)
+			} else {
+				if _, err = deleteObject(ctx, conn, &DeleteRequest{
+					Bucket:       bucket,
+					Key:          *v.Key,
+					VersionID:    *v.VersionId,
+					ForceDestroy: force,
+				}); err != nil {
+					errs = append(errs, newObjectVersionError(toString(v.Key), toString(v.VersionId), err))
+				} else {
+					objCount++
+				}
+			}
+		} else {
+			errs = append(errs, newDeleteObjectVersionError(v))
+		}
+	}
+	if err := errors.Join(errs...); err != nil {
+		return objCount, fmt.Errorf("deleting S3 bucket (%s) object versions: %w", bucket, err)
+	}
+
+	return objCount, nil
+}
+
+func toString(v *string) string {
```

Comment:
ToValueDefault from the sdk should do the same as this function

Reply:
s3-sdk-go seems not to have that.

Reply:
you can use from sdk-go-bundle  shared.ToValueDefault. shared is already added to go.mod

Reply:
sure